src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/cas2/ApplicationsController.kt

```diff
@@ -40,31 +41,41 @@ class ApplicationsController(
   private val objectMapper: ObjectMapper,
   private val offenderService: OffenderService,
   private val userService: NomisUserService,
+  private val externalUserService: ExternalUserService,
 ) : ApplicationsCas2Delegate {
 
   override fun applicationsGet(): ResponseEntity<List<ApplicationSummary>> {
-    val user = userService.getUserForRequest()
-
-    val applications = applicationService.getAllApplicationsForUser(user)
-
-    return ResponseEntity.ok(applications.map { getPersonDetailAndTransformToSummary(it, user) })
+    val authenticatedPrincipal = httpAuthService.getCas2AuthenticatedPrincipalOrThrow()
```

Comment:
do I understand it correctly then that there are broadly two places where we impose permissions:
- the `OAuth2ResourceServerSecurityConfiguration.kt` file where we impose _role-based authorization_ on paths. So, e.g. only roles `CAS2_ASSESSOR` and `PRISON` will be able to reach the cas2/applications endpoints.
- the `getCas2AuthenticatedPrincipalOrThrow` which imposes _auth_source authorization_, so e.g. we look at the auth source and if it isn't `nomis` or `auth` they are logged out?

Reply:
yes, that's how it works at present. I think that we should develop this towards:

#### 1. `OAuth2ResourceServerSecurityConfiguration`
Continue to enforce role-based access rules working on both permitted HTTP method and path. But as previously discussed use custom matchers in the Spring Security filter chain, something like:
```kotlin
val cas2ApplicantAccess: RequestMatcher = { (request) ->
  request.getServletPath().matches("cas2/applications/**")
  // ...other conditions
}

val cas2AssessorAccess: RequestMatcher = { (request) ->
  request.getServletPath().matches("cas2/submissions/**")
  // ...other conditions
}

val cas2AdminAccess: RequestMatcher = { (request) ->
  request.getServletPath().matches("cas2/applications/**")
  // ...other conditions
}

http {
  authorizeHttpRequests {
    authorize(cas1Access, hasRole("PROBATION"))
    authorize(cas2ApplicantAccess), hasAnyRole("POM_USER", "CAS2_ADMIN"))
    authorize(cas2AssessorAccess), hasRole("CAS2_ASSESSOR"))
    authorize(cas2AdminAccess, hasRole("CAS2_ADMIN"))
    authorize(cas2MangementInfoAccess, hasRole("CAS2_MI"))
    authorize(cas3Access, hasRole("PROBATION"))
  }
}
```

These custom matchers should be unit-testable, allowing us to remove slow access tests at the integration test level.

### 2. `HttpAuthService`

`HttpAuthService.getCas2AuthenticatedPrincipalOrThrow()` tests the logged in user's JWT's claim `auth_source` and now ejects anyone who hasn't been authenticated by NOMIS or externally. We use it in our controller to i) perform that control and ii) to distinguish between external NACRO and internal NOMIS/DPS users. I think it would be good to develop this into a method such as `getCurrentuser()` which is run on every controller action to provide a persisted `Cas2User` record which is built from either a `NomisUser` or an `ExternalUser` and offers a standard interface e.g.

```
type (nomis/nacro)
userIdentifier
name
email
phone
roles (POM_USER, CAS2_ADMIN)
isEnabled
```

This user could then be passed to services by the controller where more fine-grained controls are needed, or where auditing or domain events need user information.